### ISSUE TITLE
`-Eexplicitly` instead of `-E` suggested for `echo`

### ISSUE THREAD

**Issue by EmilyGraceSeville7cf** [akinomyoga/ble.sh#325]
**ble version**: 0.4.0-devel4+f4c973b
**Bash version**: 5.1.16(1)-release (x86_64-pc-linux-gnu)

![image](https://github.com/akinomyoga/ble.sh/assets/42812113/4872abc7-a900-4a45-8a3e-c7b710a6da11)

**Comment by akinomyoga** [comment]
Hmm, interesting. It doesn't reproduce in my environment. What is the output of the following command?

```bash
$ ble/widget/display-shell-version
```

**Comment by EmilySeville7cfg** [comment]
```
GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu) [Ubuntu 22.04.2 LTS]
ble.sh, version 0.4.0-devel4+f4c973b (noarch) [git 2.34.1, GNU Make 4.3, GNU Awk 5.1.0, API: 3.0 (GNU MPFR 4.1.0, GNU MP 6.2.1)]
bash-completion, version 2.11 (hash:b42f5d6a7ad6d4921ec73838ba54a96d6bd30936, 77071 bytes) (noarch)
locale: LANG=en_US.UTF-8 LC_ADDRESS=en_CA.UTF-8 LC_IDENTIFICATION=en_CA.UTF-8 LC_MEASUREMENT=en_CA.UTF-8 LC_MONETARY=en_CA.UTF-8 LC_NAME=en_CA.UTF-8 LC_NUMERIC=en_CA.UTF-8 LC_PAPER=en_CA.UTF-8 LC_TELEPHONE=en_CA.UTF-8 LC_TIME=en_CA.UTF-8
terminal: TERM=xterm-256color wcwidth=14.0-west/15.0-2+ri, vte:6800 (65;6800;1)
```

**Comment by akinomyoga** [comment]
OK, thanks. I could reproduce the behavior in both Ubuntu 20.04 and 22.04. I'll later check what's going on there.

**Comment by EmilySeville7cfg** [comment]
@akinomyoga, here are test results for some builtins:

```
alias:
- invalid options suggested, like "-pprint" instead of "-p"

bind:
- no descriptions for options displayed

cd:
- no "-" displayed for single letter options, like "L" instead of "-L"

command:
- no options suggested

declare:
- invalid options suggested, like "-Ato" instead of "-A"

typeset:
- no options suggested

dirs:
- no options suggested

disown:
- no options suggested

echo:
- invalid options suggested, like "-eenable" instead of "-e"

exec:
- no options suggested

export:
- invalid options suggested, like "-frefer" instead of "-f"

fc:
- no options suggested

hash:
- no options suggested

help:
- no options suggested

history:
- no options suggested

jobs:
- no options suggested

kill:
- not all options suggested
```

**Comment by EmilySeville7cfg** [comment]
Is there any way to extract the completion list in a parsable format to automatically check the difference between actual acceptable options and errors in complete suggestions?

**Comment by akinomyoga** [comment]
> Is there any way to extract the completion list in a parsable format to automatically check the difference between actual acceptable options and errors in complete suggestions?

If you would like to access the option list that ble.sh internally generates, you can run the following (please replace `CommandName` with an actual command name):

```bash
$ ble/complete/mandb/generate-cache CommandName && less "$ret"
```

When `ble/complete/mandb/generate-cache` succeeds, it sets a filename to the shell variable `ret`. The file is a DSV (like CSV 
 or TSV) with the separator being FS (`$'\x1c'`). The first column is the option. The second column is the display suffix which is usually used to indicate the option argument. The third column is the suffix that is appended when the full-word completion is performed. The fourth column is the description.

**Comment by EmilySeville7cfg** [comment]
@akinomyoga, here is my script:

```bash
#!/usr/bin/env bash

# shellcheck disable=1090
source ~/.local/share/blesh/ble.sh --lib
ble-import core-complete

declare -A command_options=(
    [":"]=""
    [source]=""
    [alias]="-p"
    [bg]=""
    [bind]="-m
-l
-p
-s
-v
-P
-S
-V
-X
-q
-u
-r
-f
-x"
    [break]=""
    [builtin]=""
    [caller]=""
    [cd]="-L
-P
-e
-@"
    [command]="-p
-V
-v"
    [compgen]="-a
-b
-c
-d
-e
-f
-g
-j
-k
-s
-u
-v
-o
-D
-E
-I
-A
-G
-W
-F
-C
-X
-P
-S"
    [complete]="-a
-b
-c
-d
-e
-f
-g
-j
-k
-s
-u
-v
-o
-D
-E
-I
-A
-G
-W
-F
-C
-X
-P
-S
-p
-r
"
    [compopt]="-o
-D
-E
-I
+o"
    [continue]=""
    [declare]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [typeset]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [dirs]="-c
-l
-p
-v"
    [disown]="-a
-r
-h"
    [echo]="-n
-e
-E"
    [enable]="-a
-d
-n
-p
-s
-f"
    [eval]=""
    [exec]="-c
-l
-a"
    [exit]=""
    [export]="-f
-n
-p"
    [fc]="-e
-l
-n
-r
-s"
    [fg]=""
    [getopts]=""
    [hash]="-l
-r
-p
-d
-t"
    [help]="-d
-m
-s"
    [history]="-c
-d
-a
-n
-r
-w
-p
-s"
    [jobs]="-l
-n
-p
-r
-s
-x"
    [kill]="-s
-n
-l
-L"
    [local]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [logout]=""
    [mapfile]="-d
-n
-O
-s
-t
-u
-C
-c"
    [readarray]="-d
-n
-O
-s
-t
-u
-C
-c"
    [popd]="-n"
    [printf]="-v"
    [pushd]="-n"
    [pwd]="-L
-P"
    [read]="-e
-r
-s
-a
-d
-i
-n
-N
-p
-t
-u"
    [readonly]="-a
-A
-f
-p"
    [return]=""
    [set]="-a
-b
-e
-f
-h
-k
-m
-n
-p
-t
-u
-v
-x
-B
-C
-E
-H
-P
-T
-o
+a
+b
+e
+f
+h
+k
+m
+n
+p
+t
+u
+v
+x
+B
+C
+E
+H
+P
+T
+o"
    [shift]=""
    [shopt]="-p
-q
-s
-u
-o"
    [suspend]="-f"
    [test]="-a
-b
-c
-d
-e
-f
-g
-h
-k
-p
-r
-s
-t
-u
-w
-x
-G
-L
-N
-O
-S
-ef
-nt
-ot
-o
-v
-R
-z
-n
-eq
-ne
-gt
-lt
-ge
-le"
    [times]=""
    [trap]="-l
-p"
    [type]="-a
-f
-t
-p
-P"
    [ulimit]="-H
-S
-a
-b
-c
-d
-e
-f
-i
-k
-l
-m
-n
-p
-q
-r
-s
-t
-u
-v
-x
-P
-R
-T"
    [umask]="-p
-S"
    [unalias]="-a"
    [unset]="-f
-v
-n"
    [wait]="-f
-n
-p"
)

for name in "${!command_options[@]}"; do
    echo "$name:"
    # shellcheck disable=SC2154
    ble/complete/mandb/generate-cache "$name" && {
        declare suggestions="$(awk --field-separator=$'\x1c' '{print $1}' "$ret")"
        while read -r option; do
            echo "$suggestions" | grep --quiet -- "$option" || echo "- '$option' is not suggested, while it's documented"
        done < <(echo "${command_options[$name]}")

        while read -r option; do
            echo "${command_options[$name]}" | grep --quiet -- "$option" || echo "- '$option' is suggested, while it's unknown"
        done < <(echo "$suggestions")
    }
    echo
done
```

and here's what it says:

```
hash:

unalias:

local:

type:

pushd:

unset:

exec:

pwd:
- '-Lprint' is suggested, while it's unknown
- '-Pprint' is suggested, while it's unknown
- '--help' is suggested, while it's unknown
- '--logical' is suggested, while it's unknown
- '--physical' is suggested, while it's unknown
- '--version' is suggested, while it's unknown

mapfile:

history:
- '-c' is not suggested, while it's documented
- '-d' is not suggested, while it's documented
- '-a' is not suggested, while it's documented
- '-n' is not suggested, while it's documented
- '-r' is not suggested, while it's documented
- '-w' is not suggested, while it's documented
- '-p' is not suggested, while it's documented
- '-s' is not suggested, while it's documented
- '-1' is suggested, while it's unknown

getopts:

echo:
- '-eenable' is suggested, while it's unknown
- '-Eexplicitly' is suggested, while it's unknown
- '-ndo' is suggested, while it's unknown
- '--help' is suggested, while it's unknown
- '--version' is suggested, while it's unknown

shift:

fc:

fg:

compgen:

::

return:

enable:

umask:

alias:
- '-pprint' is suggested, while it's unknown

popd:

wait:

caller:

disown:

source:

bind:

builtin:

break:

suspend:

dirs:

printf:
- '--help' is suggested, while it's unknown
- '--version' is suggested, while it's unknown

logout:

ulimit:

typeset:
- '-a' is not suggested, while it's documented
- '-A' is not suggested, while it's documented
- '-f' is not suggested, while it's documented
- '-F' is not suggested, while it's documented
- '-g' is not suggested, while it's documented
- '-i' is not suggested, while it's documented
- '-I' is not suggested, while it's documented
- '-l' is not suggested, while it's documented
- '-n' is not suggested, while it's documented
- '-r' is not suggested, while it's documented
- '-t' is not suggested, while it's documented
- '-u' is not suggested, while it's documented
- '-x' is not suggested, while it's documented
- '-p' is not suggested, while it's documented
- '+f' is not suggested, while it's documented

readonly:

cd:

times:

jobs:

kill:
- '-s' is not suggested, while it's documented
- '-n' is not suggested, while it's documented
- '--list' is suggested, while it's unknown
- '--queue' is suggested, while it's unknown
- '-q' is suggested, while it's unknown
- '-<signal>' is suggested, while it's unknown
- '--table' is suggested, while it's unknown

readarray:

eval:

shopt:

complete:
- '-Dapply' is suggested, while it's unknown
- '-Eapply' is suggested, while it's unknown
- '-Iapply' is suggested, while it's unknown
- '-pprint' is suggested, while it's unknown
- '-rremove' is suggested, while it's unknown

declare:
- '+f' is not suggested, while it's documented
- '-Ato' is suggested, while it's unknown
- '-ato' is suggested, while it's unknown
- '-frestrict' is suggested, while it's unknown
- '-Frestrict' is suggested, while it's unknown
- '-gcreate' is suggested, while it's unknown
- '-Iif' is suggested, while it's unknown
- '-ito' is suggested, while it's unknown
- '-lto' is suggested, while it's unknown
- '-nmake' is suggested, while it's unknown
- '-pdisplay' is suggested, while it's unknown
- '-rto' is suggested, while it's unknown
- '-tto' is suggested, while it's unknown
- '-uto' is suggested, while it's unknown
- '-xto' is suggested, while it's unknown

export:
- '-frefer' is suggested, while it's unknown
- '-nremove' is suggested, while it's unknown
- '-pdisplay' is suggested, while it's unknown

exit:

compopt:

continue:

trap:

bg:

help:

set:

read:

test:
- '--help' is suggested, while it's unknown
- '--version' is suggested, while it's unknown

command:

```

What about add unit tests to check whether completions for builtins work fine?

**Comment by akinomyoga** [comment]
> What about add unit tests to check whether completions for builtins work fine?

Thanks for the test cases! I'll use them to test the current implementation.

But the available sets of flags of builtins depend on the Bash version. I currently run the tests on

- bash-3.0.22, bash-3.1.23, bash-3.2.57, bash-4.0.44, bash-4.1.17, bash-4.2.53, bash-4.3.48, bash-4.4.23, bash-5.0.18, bash-5.1.16, bash-5.2.9

so one needs to maintain such a list for every Bash version. Also, we need to add the expected sets of flags every time a new Bash version appears, which means that this explicitly adds the maintenance cost that continues forever. I wouldn't like to add the future maintenance cost for it.

----

I decided to add the settings for all the builtins to extract options using the `help` builtin with adjustments. Then, I applied your script, but there are still differences. Could you check the option lists in your script for the following commands?

- `declare`, `typeset`, and `local`: You are requesting support for `+f`. Although specifying `+f` to these builtins does not cause an error, what is the effect of `+f`? Does it have an actual effect? The other accepted but no-effect options `+p`, `+I`, `+g`, and `+F` are not requested by the list.
- `compgen`: Your script requests -DEI, but compgen doesn't support them.

**Comment by EmilySeville7cfg** [comment]
> so one needs to maintain such a list for every Bash version.

We can do it in the following way:

```yaml
<bash-version>:
  <command>: [<flags>]
```

**Comment by akinomyoga** [comment]
> We can do it in the following way:

What do you mean? Of course, we can do it technically. But we *need to do it* anyway. I'm not talking about the technical difficulty but the maintenance cost.

**Comment by EmilySeville7cfg** [comment]
[Here](https://gist.github.com/EmilySeville7cfg/6ecd48b4b0ed616c77fcc80f00dfb536) are outputs of declare commands.

**Comment by akinomyoga** [comment]
> [Here](https://gist.github.com/EmilySeville7cfg/6ecd48b4b0ed616c77fcc80f00dfb536) are outputs of `declare` commands.

As far as I look at the results in the link, they seem to prove that `+f` doesn't have an effect while `-f` has an effect. Even `declare -p -f +f` seems to produce the identical result as `declare -p -f`.

> YAML allows [reusing](https://support.atlassian.com/bitbucket-cloud/docs/yaml-anchors/) it's items like this. So I guess cost is not so high.

The cost I'm talking about is not a computational cost but the fact that we need to *manually* update the list every time a new Bash version that has new options appears. Someone needs to continuously check the new Bash versions. You might suggest it to automate it, but then there is another maintenance cost to check whether the automation is not broken by the new Bash versions, which would further add the cost.

**Comment by EmilySeville7cfg** [comment]
We can partially automate it: parse man page for new bash version, generate YAML code, and check it manually. I agree, it's an additional work, but if not, then what? 🤔 I wanna be sure that at least builtins have correct completions.

**Comment by akinomyoga** [comment]
> parse man page for new bash version,

This is what the current ble.sh does (but just hasn't specifically configured for builtins).

> generate YAML code, and check it manually.

If you finally check it manually, there doesn't seem to be much difference from manually checking the result of the ble.sh's parsing. What you are suggesting seems to me that to avoid manually testing function A, prepare another function A' to do essentially the same thing, and manually test function A'. This seems to make the procedure just indirect but doesn't reduce any manual testing.

**Comment by akinomyoga** [comment]
I pushed fixes:

> `-Eexplicitly` instead of `-E` suggested for `echo` (title)

- 313cfb25 This is the fix for that the first word of the description connects to the option.

> https://github.com/akinomyoga/ble.sh/issues/325#issuecomment-1545474274
> ```
> cd:
> - no "-" displayed for single letter options, like "L" instead of "-L"
> ```

This behavior has a reason. The built-in `cd` completion experimentally implemented the completion for each character in the combined short flags, such as `-LP[here]`, etc., where only the character of the completed short flag is considered the target of the completion. The completion for the first character `-[here]` is just a special case for such a short-option completion inside the set of flags. However, it is not so clear in the completion menu list that it is a part of the short option.

- 74402098 I decided to prefix `-` in the completion menu while keeping the current behavior.

> https://github.com/akinomyoga/ble.sh/issues/325#issuecomment-1546542201

The completions for all the other builtins were just not implemented.

- 9bd24691 The completion for builtins can be generally produced by parsing the result of the command `help "$builtin_name"`, so I configured it. However, some flags are still not shown by `help "$builtin"` or are difficult to parse. The descriptions for some options are manually added (in particular, the flags for `complete` and `wait`).

----

> https://github.com/akinomyoga/ble.sh/issues/325#issuecomment-1546542201
>
> @akinomyoga, here is my script:
> 
> ```shell
> #!/usr/bin/env bash
> 
> # shellcheck disable=1090
> source ~/.local/share/blesh/ble.sh --lib
> ble-import core-complete
> 
> declare -A command_options=(
> ```

I forgot to tell you, but could you add the line `ble-import core-cmdspec` after `ble-import core-complete`? After adding it, you might also need to clear the cache by running `source /path/to/ble.sh --clear-cache`.

**Comment by EmilySeville7cfg** [comment]
Like this?

```bash
#!/usr/bin/env bash

# shellcheck disable=1090
source ~/.local/share/blesh/ble.sh --lib
ble-import core-complete
ble-import core-cmdspec

declare -A command_options=(
    [":"]=""
    [source]=""
    [alias]="-p"
    [bg]=""
    [bind]="-m
-l
-p
-s
-v
-P
-S
-V
-X
-q
-u
-r
-f
-x"
    [break]=""
    [builtin]=""
    [caller]=""
    [cd]="-L
-P
-e
-@"
    [command]="-p
-V
-v"
    [compgen]="-a
-b
-c
-d
-e
-f
-g
-j
-k
-s
-u
-v
-o
-A
-G
-W
-F
-C
-X
-P
-S"
    [complete]="-a
-b
-c
-d
-e
-f
-g
-j
-k
-s
-u
-v
-o
-D
-E
-I
-A
-G
-W
-F
-C
-X
-P
-S
-p
-r
"
    [compopt]="-o
-D
-E
-I
+o"
    [continue]=""
    [declare]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [typeset]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [dirs]="-c
-l
-p
-v"
    [disown]="-a
-r
-h"
    [echo]="-n
-e
-E"
    [enable]="-a
-d
-n
-p
-s
-f"
    [eval]=""
    [exec]="-c
-l
-a"
    [exit]=""
    [export]="-f
-n
-p"
    [fc]="-e
-l
-n
-r
-s"
    [fg]=""
    [getopts]=""
    [hash]="-l
-r
-p
-d
-t"
    [help]="-d
-m
-s"
    [history]="-c
-d
-a
-n
-r
-w
-p
-s"
    [jobs]="-l
-n
-p
-r
-s
-x"
    [kill]="-s
-n
-l
-L"
    [local]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [logout]=""
    [mapfile]="-d
-n
-O
-s
-t
-u
-C
-c"
    [readarray]="-d
-n
-O
-s
-t
-u
-C
-c"
    [popd]="-n"
    [printf]="-v"
    [pushd]="-n"
    [pwd]="-L
-P"
    [read]="-e
-r
-s
-a
-d
-i
-n
-N
-p
-t
-u"
    [readonly]="-a
-A
-f
-p"
    [return]=""
    [set]="-a
-b
-e
-f
-h
-k
-m
-n
-p
-t
-u
-v
-x
-B
-C
-E
-H
-P
-T
-o
+a
+b
+e
+f
+h
+k
+m
+n
+p
+t
+u
+v
+x
+B
+C
+E
+H
+P
+T
+o"
    [shift]=""
    [shopt]="-p
-q
-s
-u
-o"
    [suspend]="-f"
    [test]="-a
-b
-c
-d
-e
-f
-g
-h
-k
-p
-r
-s
-t
-u
-w
-x
-G
-L
-N
-O
-S
-ef
-nt
-ot
-o
-v
-R
-z
-n
-eq
-ne
-gt
-lt
-ge
-le"
    [times]=""
    [trap]="-l
-p"
    [type]="-a
-f
-t
-p
-P"
    [ulimit]="-H
-S
-a
-b
-c
-d
-e
-f
-i
-k
-l
-m
-n
-p
-q
-r
-s
-t
-u
-v
-x
-P
-R
-T"
    [umask]="-p
-S"
    [unalias]="-a"
    [unset]="-f
-v
-n"
    [wait]="-f
-n
-p"
)

declare -i counter=0

for name in "${!command_options[@]}"; do
    echo "$name:"
    # shellcheck disable=SC2154
    ble/complete/mandb/generate-cache "$name" && {
        declare suggestions="$(awk --field-separator=$'\x1c' '{print $1}' "$ret")"
        declare -i is_broken=0
        while read -r option; do
            echo "$suggestions" | grep --quiet -- "$option" || {
                echo "- '$option' is not suggested, while it's documented"
                is_broken=1
            }
        done < <(echo "${command_options[$name]}")

        while read -r option; do
            echo "${command_options[$name]}" | grep --quiet -- "$option" || {
                echo "- '$option' is suggested, while it's unknown"
                is_broken=1
            }
        done < <(echo "$suggestions")
        counter+=$is_broken
    }
    echo
done

echo "$counter out of ${#command_options[@]} are broken"
source /path/to/ble.sh --clear-cache

```

**Comment by EmilySeville7cfg** [comment]
```bash
#!/usr/bin/env bash

# shellcheck disable=1090
source ~/.local/share/blesh/ble.sh --lib
ble-import core-complete
ble-import core-cmdspec

declare -A command_options=(
    [":"]=""
    [source]=""
    [alias]="-p"
    [bg]=""
    [bind]="-m
-l
-p
-s
-v
-P
-S
-V
-X
-q
-u
-r
-f
-x"
    [break]=""
    [builtin]=""
    [caller]=""
    [cd]="-L
-P
-e
-@"
    [command]="-p
-V
-v"
    [compgen]="-a
-b
-c
-d
-e
-f
-g
-j
-k
-s
-u
-v
-o
-A
-G
-W
-F
-C
-X
-P
-S"
    [complete]="-a
-b
-c
-d
-e
-f
-g
-j
-k
-s
-u
-v
-o
-D
-E
-I
-A
-G
-W
-F
-C
-X
-P
-S
-p
-r
"
    [compopt]="-o
-D
-E
-I
+o"
    [continue]=""
    [declare]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [typeset]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [dirs]="-c
-l
-p
-v"
    [disown]="-a
-r
-h"
    [echo]="-n
-e
-E"
    [enable]="-a
-d
-n
-p
-s
-f"
    [eval]=""
    [exec]="-c
-l
-a"
    [exit]=""
    [export]="-f
-n
-p"
    [fc]="-e
-l
-n
-r
-s"
    [fg]=""
    [getopts]=""
    [hash]="-l
-r
-p
-d
-t"
    [help]="-d
-m
-s"
    [history]="-c
-d
-a
-n
-r
-w
-p
-s"
    [jobs]="-l
-n
-p
-r
-s
-x"
    [kill]="-s
-n
-l
-L"
    [local]="-a
-A
-f
-F
-g
-i
-I
-l
-n
-r
-t
-u
-x
-p
+a
+A
+f
+i
+l
+n
+r
+t
+u
+x"
    [logout]=""
    [mapfile]="-d
-n
-O
-s
-t
-u
-C
-c"
    [readarray]="-d
-n
-O
-s
-t
-u
-C
-c"
    [popd]="-n"
    [printf]="-v"
    [pushd]="-n"
    [pwd]="-L
-P"
    [read]="-e
-r
-s
-a
-d
-i
-n
-N
-p
-t
-u"
    [readonly]="-a
-A
-f
-p"
    [return]=""
    [set]="-a
-b
-e
-f
-h
-k
-m
-n
-p
-t
-u
-v
-x
-B
-C
-E
-H
-P
-T
-o
+a
+b
+e
+f
+h
+k
+m
+n
+p
+t
+u
+v
+x
+B
+C
+E
+H
+P
+T
+o"
    [shift]=""
    [shopt]="-p
-q
-s
-u
-o"
    [suspend]="-f"
    [test]="-a
-b
-c
-d
-e
-f
-g
-h
-k
-p
-r
-s
-t
-u
-w
-x
-G
-L
-N
-O
-S
-ef
-nt
-ot
-o
-v
-R
-z
-n
-eq
-ne
-gt
-lt
-ge
-le"
    [times]=""
    [trap]="-l
-p"
    [type]="-a
-f
-t
-p
-P"
    [ulimit]="-H
-S
-a
-b
-c
-d
-e
-f
-i
-k
-l
-m
-n
-p
-q
-r
-s
-t
-u
-v
-x
-P
-R
-T"
    [umask]="-p
-S"
    [unalias]="-a"
    [unset]="-f
-v
-n"
    [wait]="-f
-n
-p"
)

declare -i counter=0

for name in "${!command_options[@]}"; do
    echo "$name:"
    # shellcheck disable=SC2154
    ble/complete/mandb/generate-cache "$name" && {
        declare suggestions="$(awk --field-separator=$'\x1c' '{print $1}' "$ret")"
        declare -i is_broken=0
        while read -r option; do
            echo "$suggestions" | grep --quiet -- "$option" || {
                echo "- '$option' is not suggested, while it's documented"
                is_broken=1
            }
        done < <(echo "${command_options[$name]}")

        while read -r option; do
            echo "${command_options[$name]}" | grep --quiet -- "$option" || {
                echo "- '$option' is suggested, while it's unknown"
                is_broken=1
            }
        done < <(echo "$suggestions")
        counter+=$is_broken
    }
    echo
done

echo "$counter out of ${#command_options[@]} are broken"
source ~/.local/share/blesh/ble.sh --clear-cache
```

**Comment by akinomyoga** [comment]
@EmilySeville7cfg `+f` is still there. So are you still requesting `+f` for `declare`, `typeset`, and `local`? As I have replied in https://github.com/akinomyoga/ble.sh/issues/325#issuecomment-1546621353, `+f` doesn't seem to be an actual effect like the other combinations such as `+p`.

**Comment by EmilySeville7cfg** [comment]
Sorry, I forgot remove `+f`.

**Comment by akinomyoga** [comment]
OK! Thanks!

**Comment by akinomyoga** [comment]
It's already included in commit [`9bd24691`](https://github.com/akinomyoga/ble.sh/commit/9bd2469127563262f88b1bc2f01e898854421265#diff-91a96c0ef0156987d452f7fc54d15676b1a46868497a9b6af3e88244595cc3f2).

**Comment by akinomyoga** [comment]
@EmilyGraceSeville7cf I guess I can also close this issue?
